S3guard/hadoop 13786 committer

ehiggs opened the pull request:
Various small fixes to the Hadoop s3guard patch set.

ehiggs committed:
Refactor Pendingset to become PendingSet.

ehiggs committed:
Use hadoop-commons Pair instead of rolling our own local Pair class.

ehiggs committed:
Removing unused functions.

ehiggs committed:
Make splitPathToElements more concise.

ehiggs committed:
Do not use newArrayList as Guava use is to be reduced afaict.

ehiggs commented:
Rebased.

steveloughran commented:
heh, I've just gone and broken things by squashing patches

steveloughran commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/commit/magic/MagicS3GuardCommitter.java:
```diff
@@ -22,9 +22,11 @@
 import java.util.Collections;
 import java.util.List;
 
+import org.apache.hadoop.fs.s3a.commit.files.PendingSet;
```
don't like imports to be moved around. As well as a nominal rule of layout, changes here make merging a nightmare

steveloughran commented on hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3a_committer_architecture.md:
```diff
@@ -951,23 +951,23 @@ to the job attempt.
 1. The contents are loaded into a list of single pending uploads.
 1. These are merged into to a single `Pendingset` structure.
-1. Which is saved to a `.pendingset` file in the job attempt directory.
+1. Which is saved to a `.pendingSet` file in the job attempt directory.
```
no. it's "pendingset" lower case.

steveloughran commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/commit/CommitOperations.java:
```diff
@@ -30,9 +30,11 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import org.apache.commons.lang3.tuple.Pair;
```
never knew of this. good to use something else by other people.

steveloughran commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/commit/files/PendingSet.java:
```diff
@@ -47,8 +47,8 @@
  */
 @InterfaceAudience.Private
 @InterfaceStability.Unstable
-public class Pendingset extends PersistentCommitData {
```
Ignoring the fact that case-changing classnames is something very unlikely to take reliably in macos unless you can delete the old one first, I'm unsure about this. Why. It's not really a "set" in terms of java lang collection. At the same time, I'm not that fussy about classnames, though I do want it to stay all lower in the filesystem.

ehiggs replied:
Well it's two words. Pending and Set. So it's camel cased. Maybe `PendingBatch`?

The issue about case changing and macos doesn't make sense. The patch wasn't merged yet so no one will have issues with the name change.

steveloughran commented:
overall

+1 to the new pair class, guava stripping, etc
=0 to the pendingset/pendingSet classchange; -1 to making the filename mixed case
-1 to moving imports around; especially when it doesn't follow the (nominal) hadoop ordering

ehiggs committed:
Fix .pendingSet to be .pendingset again

ehiggs committed:
Soem grammar fixes. I think I fixed them to what was intended.

ehiggs committed:
Move PendingSet imports back to where they belong.

ehiggs commented:
I rebased and added a few patched addressing @steveloughran's comments.

I also fixed up some grammar or sentence fragments in the architecture doc.

steveloughran commented:
thanks, merged in
